File: leetcode/hashmap/dart/top_k_frequent_elements.dart
class Solution {
  List<int> topKFrequent(List<int> nums, int k) {
    final hashmap = <int, int>{};
    for (var i = 0; i < nums.length; i++) {
      hashmap[nums[i]] = (hashmap[nums[i]] ?? 0) + 1;
    }
    var result = <int>[];
    var sortedByValueMap = Map<int, int>.fromEntries(hashmap.entries.toList()
      ..sort((e1, e2) => e2.value.compareTo(e1.value)));
    sortedByValueMap.forEach((key, value) {
      if (result.length < k) {
        result.add(key);
      }
    });
    return result;
  }
}

class Solution2 {
  List<int> topKFrequent(List<int> nums, int k) {
    final hashmap = <int, int>{};
    for (int i = 0; i < nums.length; i++) {
      hashmap[nums[i]] = (hashmap[nums[i]] ?? 0) + 1;
    }
    final result = <int>[];
    for (int i = 0; i < k; i++) {
      int maxCount = 0;
      int maxCandidate = 0;
      hashmap.entries.forEach((element) {
        if (element.value > maxCount) {
          maxCandidate = element.key;
          maxCount = element.value;
        }
      });
      result.add(maxCandidate);
      hashmap.remove(maxCandidate);
    }
    return result;
  }
}

void main() {
  final nums = [1, 1, 1, 2, 2, 2, 2, 3, 3];
  final k = 3;
  final obj = Solution();
  print(obj.topKFrequent(nums, k));
}

File: leetcode/hashmap/go/top_k_frequent_elements.go
package hashmap

import "fmt"

func getMapValue(m map[int]int) []int {
	values := make([]int, 0, len(m))
	// For the map, the for loop with i1, and i2 indicates for the key, and value
	for _, v := range m {
		values = append(values, v)
	}
	return values
}

func topKFrequent(nums []int, k int) []int {
	result := make([]int, k)
	hashmap := make(map[int]int)

	for _, v := range hashmap {
		hashmap[v]++
	}
	fmt.Println(hashmap)
	for i := 0; i < k; i++ {
		maxCandidate := 0
		maxCount := 0

		for k, v := range hashmap {
			if v > maxCount {
				maxCandidate = k
				maxCount = v
			}
		}
		result[i] = maxCandidate
		delete(hashmap, maxCandidate)
	}
	return result
}

func main() {
	nums := []int{1, 1, 1, 2, 2, 3}
	k := 2
	fmt.Println(topKFrequent(nums, k))
}
